stdlib/public/Concurrency/TaskLocal.cpp

```diff
@@ -10,71 +10,160 @@
+                                             const Metadata *keyType,
+                                             TaskLocal::TaskLocalInheritance inheritance) {
+  return task->localValueGet(keyType, inheritance);
+}
+
+void swift::swift_task_localValuePop(AsyncTask *task) {
+  task->localValuePop();
+}
+
+// =============================================================================
+// ==== Initialization ---------------------------------------------------------
+
+void TaskLocal::Storage::initializeLinkParent(AsyncTask* task,
+                                              AsyncTask* parent) {
+  assert(!head && "initial task local storage was already initialized");
+  assert(parent && "parent must be provided to link to it");
+  head = TaskLocal::Item::createParentLink(task, parent);
+}
+
+TaskLocal::Item*
+TaskLocal::Item::createParentLink(AsyncTask *task, AsyncTask *parent) {
+  size_t amountToAllocate = Item::itemSize(/*valueType*/nullptr);
+  // assert(amountToAllocate % MaximumAlignment == 0); // TODO: do we need this?
+  void *allocation = swift_task_alloc(task, amountToAllocate);
+  Item *item = new(allocation) Item();
+
+  auto parentHead = parent->Local.head;
+  if (parentHead) {
+    if (parentHead->isEmpty()) {
+      switch (parentHead->getNextLinkType()) {
+        case NextLinkType::IsParent:
+          // it has no values, and just points to its parent,
+          // therefore skip also skip pointing to that parent and point
+          // to whichever parent it was pointing to as well, it may be its
+          // immediate parent, or some super-parent.
+          item->next = reinterpret_cast<uintptr_t>(parentHead->getNext()) |
+                       static_cast<uintptr_t>(NextLinkType::IsParent);
+          break;
+        case NextLinkType::IsNext:
+          assert(false &&
+                 "empty taskValue head in parent task, yet parent's 'head' is `IsNext`, "
+                 "this should not happen, as it implies the parent must have stored some value.");
+          break;
+        case NextLinkType::IsTerminal:
+          item->next = reinterpret_cast<uintptr_t>(parentHead->getNext()) |
+                       static_cast<uintptr_t>(NextLinkType::IsTerminal);
+          break;
+      }
+    } else {
+      item->next = reinterpret_cast<uintptr_t>(parentHead) |
+                   static_cast<uintptr_t>(NextLinkType::IsParent);
+    }
+  } else {
+    item->next = reinterpret_cast<uintptr_t>(parentHead) |
+                 static_cast<uintptr_t>(NextLinkType::IsTerminal);
+  }
+
+  return item;
+}
+
+TaskLocal::Item*
+TaskLocal::Item::createLink(AsyncTask *task,
+                        const Metadata *keyType,
+                        const Metadata *valueType) {
+  assert(task);
+  size_t amountToAllocate = Item::itemSize(valueType);
+  // assert(amountToAllocate % MaximumAlignment == 0); // TODO: do we need this?
+  void *allocation = swift_task_alloc(task, amountToAllocate);
```

Comment:
hooray, we're now using swift_task_alloc/swift_task_dealloc, and it all works excellently thanks to the only way to bind a value being `withLocal` 👍